pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/array/ListAggDistinctFunction.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.aggregation.function.array;
+
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import java.util.Collection;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.core.query.aggregation.AggregationResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+
+
+public class ListAggDistinctFunction extends ListAggFunction {
+
+  public ListAggDistinctFunction(ExpressionContext expression, String separator, boolean nullHandlingEnabled) {
+    super(expression, separator, nullHandlingEnabled);
+  }
+
+  @Override
+  protected ObjectArrayList<String> getObjectArrayList(AggregationResultHolder aggregationResultHolder) {
+    ObjectArrayListHashSet<String> valueSet = aggregationResultHolder.getResult();
+    if (valueSet == null) {
+      valueSet = new ObjectArrayListHashSet<>();
+      aggregationResultHolder.setValue(valueSet);
+    }
+    return valueSet;
+  }
+
+  @Override
+  protected ObjectArrayList<String> getObjectArrayList(GroupByResultHolder groupByResultHolder, int groupKey) {
+    ObjectArrayListHashSet<String> valueSet = groupByResultHolder.getResult(groupKey);
+    if (valueSet == null) {
+      valueSet = new ObjectArrayListHashSet<>();
+      groupByResultHolder.setValueForKey(groupKey, valueSet);
+    }
+    return valueSet;
+  }
+
+  /**
+   * Custom implementation of {@link ObjectArrayList} that maintains an extra {@link ObjectOpenHashSet} to ensure that
+   * the elements are unique.
+   * Note that this implementation only overrides the methods used in {@link ListAggFunction}, so the uniqueness check
+   * is not guaranteed, use with your own risk.
+   *
+   * @param <T> Type of the element
+   */
+  static class ObjectArrayListHashSet<T> extends ObjectArrayList<T> {
```

Comment:
This is essentially an `ObjectLinkedOpenHashSet`. Can we use it as the intermediate result instead?

Reply:
Replaced ObjectArrayListHashSet with ObjectLinkedOpenHashSet